OCPP 1.6 native stack

ssekar15 opened the pull request:
Add a support for Open Charge Point Protocol(OCPP v1.6) native stack
    
Basic support for open charge point protocol v1.6 as native stack with
below functionality
    
    1. Framework for ocpp stack and central system communication using RPC over
       websocket according to occp-j.
    2. Core profile with basic PDU

ssekar15 committed:
net: ocpp: Add a support for Open Charge Point Protocol(OCPP v1.6) stack

Basic support for open charge point protocol v1.6 as native stack with
below functionality

1. Framework for ocpp stack and central system communication using RPC over
   websocket according to occp-j.
2. Core profile with basic PDU

Signed-off-by: Saravanan Sekar <saravanan@linumiz.com>
Co-authored-by: Sanjay Vallimanalan <sanjay@linumiz.com>

ssekar15 committed:
samples: net: ocpp: Charge Point test sample

A test sample for ocpp charge point with 2 connectors, central
system steve server is used with test setup (public available)

Signed-off-by: Saravanan Sekar <saravanan@linumiz.com>

ssekar15 committed:
doc: ocpp: Describe Open Charge Point Protocol (OCPP) support

Describe OCPP Charge Point (CP) support and illustrate how to implement
CP sample.

Signed-off-by: Saravanan Sekar <saravanan@linumiz.com>

jukkar commented on include/zephyr/net/ocpp.h:
```diff
@@ -0,0 +1,253 @@
+ */
+
+#ifndef ZEPHYR_INCLUDE_NET_OCPP_H
+#define ZEPHYR_INCLUDE_NET_OCPP_H
```
All the header files should follow similar naming for guard defines so this should be
```
#ifndef ZEPHYR_INCLUDE_NET_OCPP_H_
```
so you should add `_` at the end. Not sure if there is some tool that checks these.

jukkar commented on include/zephyr/net/ocpp.h:
```diff
@@ -0,0 +1,253 @@
+	AUTH_ACCEPTED,
+	AUTH_BLOCKED,
+	AUTH_EXPIRED,
+	AUTH_INVALID,
```
If the values are not fixed, it would be nice if the AUTH_INVALID would be first so that static status value would get 0 (invalid) values as a default one.
Also you need to name space the enum values, just add `OCPP_...` prefix.

We usually only use typedefs for function pointers in Zephyr, so I would suggest you remove those in this file (unless it is a function pointer you are defining).

jukkar commented on include/zephyr/net/ocpp.h:
```diff
@@ -0,0 +1,253 @@
+ *         library
+ */
+typedef enum {
+	OMM_CURRENT_FROM_EV,		// in A
```
Please prefix with OCPP_

jukkar commented on include/zephyr/net/ocpp.h:
```diff
@@ -0,0 +1,253 @@
+		/** input to user, requested connector_id or 0 - main meter */
+		int id_con;
+
+		ocpp_meter_measurand_t mes;
```
All the variables should be documented by default. If some variables are kind of "private", you can add documentation guards `/** @cond INTERNAL_HIDDEN */` to prevent them from showing in the documentation, see examples in other header files.

jukkar commented on lib/ocpp/Kconfig:
```diff
@@ -0,0 +1,74 @@
+config OCPP
+	bool "Open Charge Point Protocol"
+	depends on JSONC
+	help
```
If this depends on websocket, you should add dependency here

jukkar replied:
We have networking libraries in `subsys/net/lib` directory, should this one placed there?

jukkar commented on lib/ocpp/Kconfig:
```diff
@@ -0,0 +1,74 @@
+
+config OCPP_PROFILE_REMOTE_TRIG
+	bool "OCPP profile remote trigger"
+	default n
```
Options are "off" by default, so this line can be omitted (check the other options with "default n" too).

jukkar commented on lib/ocpp/include/ocpp_i.h:
```diff
@@ -0,0 +1,195 @@
+/*
```
You have only one file in include directory which looks quite pointless, also the src directory has couple of files. I suggest putting all the files to `ocpp/` directory.

jukkar commented on lib/ocpp/src/core.c:
```diff
@@ -0,0 +1,510 @@
+		resp = "Unlocked";
+	}
+
+	fn = ctx->cfn[UnlockConnector];
+	fn(buf, sizeof(buf), resp, uid);
```
Using a fn variable seems pointless just for this one call, why not just do `ctx->cfn[UnlockConnector](buf, ....);`

ssekar15 replied:
made for better code readability

jukkar commented on lib/ocpp/src/core.c:
```diff
@@ -0,0 +1,510 @@
+	ocpp_msg_fp_t fn;
+	int ret;
+
+	//get time stamp
+	// if msg not empty or offline add to q
```
No // comments are allowed in zephyr

jukkar commented on lib/ocpp/src/ocpp.c:
```diff
@@ -0,0 +1,747 @@
+#include "ocpp_i.h"
+#include <zephyr/posix/time.h>
+
+LOG_MODULE_REGISTER(ocpp, LOG_LEVEL_INF);
```
You should allow user to set the log level via Kconfig, see examples in other network libraries

jukkar commented on lib/ocpp/src/ocpp.c:
```diff
@@ -0,0 +1,747 @@
+#define OCPP_UPSTREAM_PRIORITY          7
+#define FILL_PDU_TABLE(_pdu)  [_pdu] = {_pdu, #_pdu, }
+
+K_THREAD_STACK_DEFINE(ocpp_int_handler_stack, CONFIG_OCPP_INT_THREAD_STACKSIZE);
+K_THREAD_STACK_DEFINE(ocpp_wsreader_stack, CONFIG_OCPP_WSREADER_THREAD_STACKSIZE);
```
These should probably be static?

jukkar reviewed (commented):
Some initial comments, will continue later.

jukkar commented:
If this PR is still a draft (as indicated by the title), you can mark it as a draft in github. Search "Still in progress?" and there is a link at top right hand side of the page.

rlubos commented on lib/ocpp/Kconfig:
```diff
@@ -0,0 +1,74 @@
+# Copyright (c) 2023 Linumiz
+# SPDX-License-Identifier: Apache-2.0
+
+config OCPP
```
IMO this should be marked as Experimental, given it relies on our Websocket library (which is still considered as an experimental).

rlubos commented on lib/ocpp/include/ocpp_i.h:
```diff
@@ -0,0 +1,195 @@
+#define to_session(ptr) CONTAINER_OF(ptr, ocpp_session_t, node)
+
+typedef enum {
+	BootNotification,
```
PascalCase shouldn't be used in Zephyr.

rlubos reviewed (commented):
Some initial thoughts. I think such a library should get a separate entry in CODEOWNERS/Maintainers file - IMO it's way too specialized to fall under general "networking" umbrella.

ssekar15 committed:
MAINTAINERS: Add Open Charge Point Protocol (OCPP) lib

add code owner for OCPP native stack

Signed-off-by: Saravanan Sekar <saravanan@linumiz.com>

ssekar15 commented:
depends on https://github.com/zephyrproject-rtos/zephyr/pull/66763

ssekar15 committed:
tests: net: ocpp: Add test suit for OCPP lib

Add test suit for OCPP lib with one complete charging session
with a random meter values.

Signed-off-by: Saravanan Sekar <saravanan@linumiz.com>

rlubos commented on subsys/net/lib/ocpp/Kconfig:
```diff
@@ -0,0 +1,76 @@
+config OCPP
+	bool "Open Charge Point Protocol [EXPERIMENTAL]"
+	depends on JSONC
+	depends on WEBSOCKET_CLIENT
```
`select EXPERIMENTAL`

rlubos commented on subsys/net/lib/ocpp/Kconfig:
```diff
@@ -0,0 +1,76 @@
+	bool "PDU message format as JSON"
+	default y
+	help
+	   If y, then the Charge Point and Central System PDU message format
```
`If enabled,`

rlubos commented on subsys/net/lib/ocpp/Kconfig:
```diff
@@ -0,0 +1,76 @@
+config OCPP_PROFILE_SMART_CHARGE
+	bool "OCPP profile smart charging"
+	help
+	  choose to y to enable smart charging profile
```
IMO this help message (and those below) are redundant with the config description, it's self-explanatory that setting config to y enables it. Either drop those help sections, or elaborate more about particular configs.

rlubos commented on subsys/net/lib/ocpp/ocpp.c:
```diff
@@ -0,0 +1,739 @@
+	if (!hndl)
+		return -EINVAL;
+
+	sh = (ocpp_session_t *)k_calloc(1, sizeof(ocpp_session_t));
```
We tend to avoid heap usage in the networking libraries unless it's really unavoidable. Wouldn't it make sense to allow the app to preallocate the session structure, and provide the pointer as an argument to the function?

ssekar15 replied:
ocpp_session_handle_t exposed to user is opaque with void *, otherwise struct ocpp_session has to be made public has app has nothing with internal members also app should care about allocate/static memory. I would prefer to keep it, suggestion please

rlubos commented on subsys/net/lib/ocpp/ocpp.c:
```diff
@@ -0,0 +1,739 @@
+		return -EINVAL;
+
+	sh = (ocpp_session_t *)k_calloc(1, sizeof(ocpp_session_t));
+	if (!sh)
```
`{ }` are required even for oneliners.

jukkar commented on samples/net/ocpp/Kconfig:
```diff
@@ -0,0 +1,22 @@
+# SPDX-License-Identifier: Apache-2.0
+
+mainmenu "OCPP sample application"
+config OCPP_SERVER
```
I would prefer what we namespace the sample specific Kconfig options properly, otherwise it is difficult to say what option is "generic" and what is sample specific. So for networking samples we have used `NET_SAMPLE` prefix for this.
So please change the options to be called `NET_SAMPLE_OCPP_SERVER`, `NET_SAMPLE_OCPP_PORT`, `NET_SAMPLE_SNTP_SERVER` etc.

jukkar commented on samples/net/ocpp/prj.conf:
```diff
@@ -0,0 +1,46 @@
+CONFIG_ETH_DRIVER=y
+CONFIG_LOG=y
+CONFIG_NET_LOG=y
+CONFIG_LOG_DEFAULT_LEVEL=3
```
I agree with @rlubos, user can always set the default log level too needed value when compiling the app.

jukkar commented on samples/net/ocpp/prj.conf:
```diff
@@ -0,0 +1,46 @@
+
+# OCPP
+CONFIG_OCPP=y
+CONFIG_OCPP_SERVER="ssh.linumiz.com"
+CONFIG_OCPP_PORT=8180
+
+CONFIG_SNTP_SERVER="162.159.200.1"
```
Also here, everything specific to your own installation should not be mentioned here. Just leave the server and sntp server empty or commented out. You can always set these separately when compiling using west.

ssekar15 replied:
These values are public reachable, anyone can directly run sample app without any changes. Please suggest if you still consider to remove it

jukkar replied:
No need to remove the options, just set empty values something like this.
```
# Please set the server addresses when compiling the sample
CONFIG_SNTP_SERVER=""
CONFIG_OCPP_SERVER=""
CONFIG_OCPP_PORT=8180
```

jukkar commented on samples/net/ocpp/src/main.c:
```diff
@@ -0,0 +1,384 @@
+	}
+
+	LOG_INF("Your address: %s",
+			net_addr_ntop(AF_INET,
+				&iface->config.dhcpv4.requested_ip,
+				buf, sizeof(buf)));
```
Indentation looks weird here and below

jukkar commented on samples/net/ocpp/src/main.c:
```diff
@@ -0,0 +1,384 @@
+	struct net_if *iface;
+
+	net_mgmt_init_event_callback(&ocpp_dhcp_cb, ocpp_handler_cb,
+			NET_EVENT_IPV4_DHCP_BOUND);
```
here too, the `NET_EVENT...` should start from the same column as `&ocpp` in previous line.

jukkar commented on samples/net/ocpp/src/main.c:
```diff
@@ -0,0 +1,384 @@
+
+			wh++;
+			LOG_DBG("mtr reading val %s con %d",
+					io->meter_val.val,
+					io->meter_val.id_con);
```
indentation

jukkar commented on samples/net/ocpp/src/main.c:
```diff
@@ -0,0 +1,384 @@
+
+		if (!tid[idx]) {
+			LOG_INF("Remote start charging idtag %s connector %d\n",
+					idtag[idx], idx + 1);
```
ditto

jukkar commented on samples/net/ocpp/src/main.c:
```diff
@@ -0,0 +1,384 @@
+				sizeof(idtag[0]));
+
+			tid[idx] = k_thread_create(&tinfo[idx], cp_stk[idx],
+						sizeof(cp_stk[idx]), ocpp_cp_entry,
+						(void *)(idx + 1), idtag[idx],
+						obs[idx], 7, 0, K_NO_WAIT);
```
ditto

jukkar commented on samples/net/ocpp/src/main.c:
```diff
@@ -0,0 +1,384 @@
+	}
+
+	LOG_INF("ocpp stop charging connector id %d\n", idcon);
+	k_sleep(K_SECONDS(1));
```
These sleeps here and below looks a bit problematic. They probably work ok now but could fail in the future if something changes in the system. If you need to wait something to finalize, better would be to have a semaphore that we wait here. We have used that kind of pattern in other networking samples.

jukkar commented on samples/net/ocpp/src/main.c:
```diff
@@ -0,0 +1,384 @@
+			break;
+		}
+
+		LOG_INF("error: ai_addrlen = %u should be %u or %u",
```
Perhaps `LOG_ERR` if this is an error

jukkar commented on samples/net/ocpp/src/main.c:
```diff
@@ -0,0 +1,384 @@
+	/* wait for device dhcp ip recive */
+	k_sleep(K_SECONDS(3));
+	ret = ocpp_getaddrinfo(CONFIG_OCPP_SERVER, CONFIG_OCPP_PORT, &ip);
+	if (ret < 0)
```
you need to use `{}` even for one liners in zephyr

jukkar commented on samples/net/ocpp/src/main.c:
```diff
@@ -0,0 +1,384 @@
+	}
+
+	/* Active charging session */
+	k_sleep(K_SECONDS(30));
```
I am a bit baffled with these sleeps, so this sample works for 30sec, stops and then sleeps 1200sec. Is there some reason for this behavior?

ssekar15 replied:
Added comments for reason

jukkar commented on samples/net/ocpp/README.rst:
```diff
@@ -0,0 +1,65 @@
+************
+
+- Linux machine
+- STEM32 Discovery kit (32F769IDISCOVERY)
```
Is this really hard requirement, or are we just depend on a board that has Ethernet?

ssekar15 replied:
No, but tested on this board. Updated doc

ssekar15 commented on include/zephyr/net/ocpp.h:
```diff
@@ -0,0 +1,254 @@
+ * @brief OCPP IdTage authorization status in result to ocpp request
+ *        authorization
+ */
+typedef enum {
```
done

ssekar15 commented:
@jukkar @rlubos Thanks for your time to review. Addressed the review comments, please consider for re-review

jukkar commented:
@ssekar15 sorry for late reply, the code looks better but could you try to fix the CI issues before continuing the reviews. I feel that some of the review issues would be resolved if the CI would pass.

ssekar15 commented:
> @ssekar15 sorry for late reply, the code looks better but could you try to fix the CI issues before continuing the reviews. I feel that some of the review issues would be resolved if the CI would pass.

@jukkar Thank you.. CI build issues are due to dependency https://github.com/zephyrproject-rtos/zephyr/pull/66763, I tried to pull one commit to check CI failure has fixed, unfortunately not. I will check on it, otherwise IMHO review shall be continued

kartben commented:
@ssekar15 will you be able to address the pending comments?

ssekar15 commented:
> @ssekar15 will you be able to address the pending comments?

@kartben Thank you for the review, yes I'm working on it

ssekar15 commented on subsys/net/lib/ocpp/core.c:
```diff
@@ -0,0 +1,503 @@
+			   struct ocpp_cp_info *cpi)
+{
+	int ret;
+	char buf[512];
```
buf is kept in context for pdu message handled from internal thread.

jukkar commented on subsys/net/lib/ocpp/core.c:
```diff
@@ -0,0 +1,504 @@
+#define FILL_METER_TABLE(_mes, _smes, _unit)[_mes] = {.mes = _mes,	\
+						 .smes = _smes, .unit = _unit}
+
+static struct {
```
Could this be const so that this struct could be placed in ROM?

jukkar commented on subsys/net/lib/ocpp/core.c:
```diff
@@ -0,0 +1,504 @@
+
+static struct {
+	enum ocpp_meter_measurand	mes;
+	char *smes;
+	char *unit;
```
These seem to be able to be `const`

jukkar commented on subsys/net/lib/ocpp/core.c:
```diff
@@ -0,0 +1,504 @@
+	union ocpp_keyval kval;
+	ocpp_msg_fp_t fn;
+	enum ocpp_key key;
+	char *res = "Accepted";
```
Perhaps const?

jukkar commented on subsys/net/lib/ocpp/core.c:
```diff
@@ -0,0 +1,504 @@
+		} else {
+			kval.str = sval;
+		}
+		ret = ocpp_update_cfg_val(key, &kval);
```
MInor style nit, please place empty line after `}` and remove the empty line between variable setting and the `if` statement

jukkar commented on subsys/net/lib/ocpp/core.c:
```diff
@@ -0,0 +1,504 @@
+		res = "NotSupported";
+	}
+
+	if (!ret) {
```
These was some rule (Misra?) that says we should only use boolean operator for boolean types. So here we should have `if (ret == 0) {` instead. This is not a big deal atm, but avoids future fixes if we done this right now.

jukkar commented on subsys/net/lib/ocpp/core.c:
```diff
@@ -0,0 +1,504 @@
+	struct ocpp_wamp_rpc_msg rmsg = {0};
+	ocpp_msg_fp_t fn;
+
+	if (!idtag || !status ||
+	    !ocpp_session_is_valid(sh)) {
```
These could be placed in the same line as it improves readability and the line length is short enough.

jukkar commented on subsys/net/lib/ocpp/core.c:
```diff
@@ -0,0 +1,504 @@
+	}
+
+	val = ocpp_get_key_val(CFG_AUTH_CACHE_ENABLED);
+	if (val && val->ival) {
+
+	}
```
Something missing here?

jukkar commented on subsys/net/lib/ocpp/core.c:
```diff
@@ -0,0 +1,504 @@
+	struct ocpp_session *sh = (struct ocpp_session *)hndl;
+	struct ocpp_info *ctx;
+	struct ocpp_upstream_info *ui;
+	char buf[350];
```
I am a bit worried about these hard coded temp buffers from stack. You have set the stack size quite large which helps here. Why do we need such a large buffer, and could we have a `define` for the size instead of hard coding the value.

jukkar commented on subsys/net/lib/ocpp/core.c:
```diff
@@ -0,0 +1,504 @@
+	}
+
+	sh->is_active = false;
+	if (atomic_dec(&ctx->mtr_timer_ref_cnt) <= 1) {
+		k_timer_stop(&ctx->mtr_timer);
+	}
```
For reference counting you should probably use `atomic_cas()` function instead.  Even if `atomic_dec` is atomic, the check after it is not. It would also make sense to create a separate ref/unref functions that would deal all the nasty details of reference counting.
See for example `subsys/net/ip/net_pkt.c` and `net_pkt_unref()` and `net_pkt_ref()` how it is done there.

ssekar15 replied:
the logical comparison of ref count is done after atomic_cas in subsys/net/ip/net_pkt.c, the same is used here atomic_dec. I don't find the if (ref <= 1) can be achieved using atomic_cas

 544         } while (!atomic_cas(&pkt->atomic_ref, ref, ref - 1));
 545 

 552         if (ref > 1) {
 553                 goto done;
 554         }

jukkar commented on subsys/net/lib/ocpp/core.c:
```diff
@@ -0,0 +1,504 @@
+				  char *uid)
+{
+	char *buf = ctx->pdu_buf;
+	char *resp = "Rejected";
```
Perhaps const

jukkar commented on subsys/net/lib/ocpp/core.c:
```diff
@@ -0,0 +1,504 @@
+				 int idtxn, char *uid)
+{
+	char *buf = ctx->pdu_buf;
+	char *resp = "Rejected";
```
ditto

jukkar commented on subsys/net/lib/ocpp/ocpp.c:
```diff
@@ -0,0 +1,748 @@
+	}
+
+	if (ui->wssock < 0) {
+		char buf[128];
```
This is a big no-no, the temp buffer is used by the websocket library for storing temporary data, if you use a variable from stack, then future websocket operations will overwrite whatever data there is in that memory.

ssekar15 replied:
As far I have checked, the API call is synchronous and host buf is copied in the below websocket (http) layer. Temp buffer are dynamic allocated, please correct me if I missed something.

jukkar replied:
Hmm, ok, I seem to have missed the fact that the buf was only used for hostname, which does not seem to cause any issues atm. It is just that the host pointer will point to invalid memory after exiting this block but fortunately the host is only used synchronously as you said. Never mind then.

jukkar commented on subsys/net/lib/ocpp/ocpp.c:
```diff
@@ -0,0 +1,748 @@
+
+	LOG_INF("upstream init");
+
+	ui = k_calloc(1, sizeof(struct ocpp_upstream_info));
```
You probably have reasons for doing dynamic memory allocations but we have been trying to avoid these (which has not always been possible) in networking code. If you do not do dynamic memory allocations you cannot have memory leaks. You could still have resource leaks but that is another issue.
Anyway, not a deal breaker but please rethink would it be possible to avoid using malloc and friends.

jukkar commented:
@ssekar15 are you planning to work on this PR?

ssekar15 commented:
> omments re: documentation have been ad

@jukkar yes for sure... looking for json dependency PR or alternative

kartben commented:
@ssekar15 I would suggest putting this PR In draft mode if it's not ready yet. This makes it clearer that it doesn't require attention from reviewers. Thanks!

ssekar15 commented:
@jukkar PR is now ready for review
Removed the external json-c module dependency and used the Zephyr json library.

subsys/net/lib/ocpp/ocpp_j.c
subsys/net/lib/ocpp/ocpp_j.h

 has the major changes

ssekar15 commented:
> Ok, I think we could try to get this to 4.3. I will review this more thoroughly after 4.2 is released. Meanwhile, could you check the CI failures?

Thanks, fixed CI failures.

rlubos reviewed (commented):
I'm not familiar with the protocol, but the library looks pretty decent. I think we should cover the library and sample in `MAINTAINERS.yml` - it would be good someone who knows the protocol have an eye on it.

ssekar15 commented on tests/net/lib/ocpp/src/main.c:
```diff
@@ -0,0 +1,121 @@
+	return 0;
+}
+
+ZTEST_EXPECT_FAIL(net_ocpp, test_ocpp_chargepoint);
```
we encounter the twister error on eth_tap and network connectivity failure, so marked as expect_fail.

```
-- west flash: using runner native
WARNING: Using a test - not safe - entropy source
[00:00:00.000,000] <err> eth_tap: Cannot create zeth (-1/Not owner)
```

rlubos replied:
So the tests are not self-contained, i.e. require the actual OCPP server to run? If that's the case, I think a better approach would be to set `build_only: true` option in testcase.yml so that twister only builds it w/o trying to execute tests.

ssekar15 commented on tests/net/lib/ocpp/src/main.c:
```diff
@@ -0,0 +1,121 @@
+	zassert_equal(ret, 0, "start transaction fail");
+
+	/* Active charging session */
+	k_sleep(K_SECONDS(20));
```
during active session, server will fetch the charging information like power delivered, voltage and current periodically in the interval of seconds. it shall be reduced/optimized for 10?

rlubos reviewed (commented):
Some final suggestions after taking a closer look at the socket handling.

ssekar15 commented on subsys/net/lib/ocpp/ocpp.c:
```diff
@@ -0,0 +1,741 @@
+	ui->csi.port = csi->port;
+	ui->csi.sa_family = csi->sa_family;
+	ui->tcpsock = zsock_socket(csi->sa_family, SOCK_STREAM,
+				   IPPROTO_TCP);
```
thanks for the review, done.

jukkar commented on subsys/net/lib/ocpp/ocpp_j.c:
```diff
@@ -0,0 +1,883 @@
+		break;
+
+	case OCPP_WAMP_RPC_ERR:
+		/* fall through */
```
ditto

jukkar commented on subsys/net/lib/ocpp/ocpp_j.c:
```diff
@@ -0,0 +1,883 @@
+
+static int parse_authorize_msg(char *json, struct ocpp_idtag_info *idtag_info)
+{
+	int ret = 0;
```
No need to initialize this as we are setting the value below

jukkar commented on subsys/net/lib/ocpp/ocpp_j.c:
```diff
@@ -0,0 +1,883 @@
+
+static int parse_heartbeat_msg(char *json, struct timeval *date)
+{
+	int ret = 0;
```
no need to initialize this

jukkar commented on subsys/net/lib/ocpp/ocpp_j.c:
```diff
@@ -0,0 +1,883 @@
+
+static int parse_idtag_info(char *json, struct ocpp_idtag_info *idtag_info)
+{
+	int ret = 0;
```
and here too, could you check all of these and remove unnecessary initializations

jukkar commented on subsys/net/lib/ocpp/ocpp_j.c:
```diff
@@ -0,0 +1,883 @@
+		return ret;
+	}
+
+	if (!payload.val1 || !payload.val2) {
```
The code is full of these kind of checks and we are trying to get rid of them because of some MISRA rule. As this is new code we should avoid introducing code that we need to change later anyway.
Could you try to change this here (and also in other parts of the code) to explicit checks like this here 
```
if (payload.val1 == 0 || payload.val2 == 0) {`
```

jukkar commented on MAINTAINERS.yml:
```diff
@@ -3717,6 +3721,17 @@ Networking:
+"Networking: OCPP":
+  status: maintained
+  maintainers:
+    - ssekar15
```
I am not sure if this is accepted, usually new users are not promoted to maintainers directly, but lets see what happens.

ssekar15 replied:
To my understanding this is accepted for new code base (new driver, or subsys or soc), and I'm already exist on MAINTAINERS.yml file

jukkar replied:
> I'm already exist on MAINTAINERS.yml file

Ah, in that case there should be no problems.

ssekar15 commented on subsys/net/lib/ocpp/ocpp.c:
```diff
@@ -0,0 +1,754 @@
+		net_sin6(addr)->sin6_port = htons(ui->csi.port);
+		zsock_inet_pton(addr->sa_family, ui->csi.cs_ip,
+				&net_sin6(addr)->sin6_addr);
+#endif
```
done

kartben commented on include/zephyr/net/ocpp.h:
```diff
@@ -0,0 +1,257 @@
+ *        authorization
+ */
+enum ocpp_auth_status {
+	OCPP_AUTH_INVALID,	/*< IdTag not valid */
```
```suggestion
	OCPP_AUTH_INVALID,	/**< IdTag not valid */
```

Fix everywhere, this is causing docs to be missing

https://builds.zephyrproject.io/zephyr/pr/68739/docs/doxygen/html/group__ocpp__api.html#ga4b2b8885d144627d43e724a3889ee1cb

kartben commented on include/zephyr/net/ocpp.h:
```diff
@@ -0,0 +1,257 @@
+ * @param[in] hndl session handle
+ * @param[in] idtag (string literal) to get authorize validity
+ * @param[out] status authorization status
+ * @param[in] timeout_ms in msec
```
```suggestion
 * @param[in] timeout_ms timeout in msec
```

kartben commented on include/zephyr/net/ocpp.h:
```diff
@@ -0,0 +1,257 @@
+ * @brief Notify transaction stopped to CS
+ *
+ * @param[in] hndl session handle
+ * @param[in] Wh energy meter reading of this connector
```
Probably better to give a more meaningful name and not have something that starts with uppercase? And explicitly mention the unit in the description 
Fix everywhere this parameter appears

kartben commented on samples/net/ocpp/README.rst:
```diff
@@ -0,0 +1,69 @@
+************
+
+- Linux machine
+- STEM32 Discovery kit (32F769IDISCOVERY) or any network interface device
```
```suggestion
- STM32 Discovery kit (32F769IDISCOVERY) or any network interface device
```

kartben reviewed (commented):
Currently on vacation so not making any of my request for changes blocking. Did only a quick pass and I think Doxygen comments could use some additional love but this can be addressed later.

ssekar15 commented on samples/net/ocpp/README.rst:
other review comments are fixed, I will leave sample part untouched as of now.

carlescufi commented:
Thanks for your perseverance and patience @ssekar15.

ssekar15 commented:
> > Thanks for your perseverance and patience @ssekar15.
> 
> +1 I will definitely reach out to get help if needed to set up some cool demos for when we will announce Zephyr 4.3!

Sure with pleasure
